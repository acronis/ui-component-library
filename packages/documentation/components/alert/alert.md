---
title: ACV Alert component
lang: en-US
editLink: true
description: This file is generated automatically from the source code. Changes made here will be lost.
---

# ACV Alert

> Alert displays a brief, important message without interrupting user's task.

<!--@include: ./alert.doc.md-->

## Props

| Prop name   | Description                                                | Type            | Values | Default |
| ----------- | ---------------------------------------------------------- | --------------- | ------ | ------- |
| description | Descriptive text. Can also be passed with the default slot | string          | -      |         |
| showBorder  | Displays border                                            | boolean         | -      |         |
| showClose   | Is close icon visible                                      | boolean         | -      |         |
| showIcon    | If a variant icon is displayed                             | boolean         | -      |         |
| title       | Title, can also be passed with the slot                    | string          | -      |         |
| subtitle    | Subtitle, can also be passed with the slot                 | string          | -      |         |
| color       | Alert variant                                              | AcvAlertVariant | -      |         |

## Events

| Event name | Properties | Description                               |
| ---------- | ---------- | ----------------------------------------- |
| close      |            | Emitted when the close button is clicked. |

## Slots

| Name        | Description                   | Bindings |
| ----------- | ----------------------------- | -------- |
| icon        | Content for the AcvAlert icon |          |
| title       | Title                         |          |
| description | Subtitle                      |          |
| default     | Custom content                |          |
| actions     | Content for the actions       |          |
| right       | Content for the right block   |          |
