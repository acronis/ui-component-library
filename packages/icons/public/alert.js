export { render as IconAlertTray16 } from '../vue/IconAlertTray16.js';
