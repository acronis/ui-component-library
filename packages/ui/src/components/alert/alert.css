@layer acv-base {
  :root {
    --acv-alert-border-radius: var(--acv-radius-medium);
    --acv-alert-title-color: var(--acv-color-text-primary);
    --acv-lert-plain-transition: 0.2s opacity
      var(--acv-transition-timing-common);
  }
}
