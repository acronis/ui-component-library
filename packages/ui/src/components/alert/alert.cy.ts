import AlertBasic from '../../../../documentation/demos/alert/AlertBasic.vue';
import Alert from './alert.vue';

describe('Alert', () => {
  it('render', () => {
    cy.mount(Alert, {
      propsData: {
        title: 'my header',
      },
    });

    cy.get('.title').should('have.text', 'my header');

    cy.matchImageSnapshot('alert-render');
  });

  it('basic', () => {
    cy.mount(AlertBasic);

    cy.matchImageSnapshot('alert-basic');
  });
});
