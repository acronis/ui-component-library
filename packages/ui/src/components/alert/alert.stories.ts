import type { Meta, StoryObj } from '@storybook/vue3';
import { ALERT_VARIANT } from '@/components/index.ts';
import AcvButton from '../button/button.vue';
import AcvAlert from './alert.vue';

type Story = StoryObj<typeof AcvAlert>;
type Args = NonNullable<Story['args']>;

export default {
  component: AcvAlert,
  title: 'Components/AcvAlert',
  tags: ['autodocs'],
  render,
  args: {
    title: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit and do it.',
    subtitle: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit and do it.',
    description: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit and do it.',
    showIcon: true,
    showClose: true
  },
} as Meta;

export const Default: Story = {};

export const NoBorder: Story = {
  args: {
    showBorder: false
  },
};

export const NoClose: Story = {
  args: {
    showClose: false
  },
};

export const NoIcon: Story = {
  args: {
    showIcon: false
  },
};

export const Info: Story = {
  args: {
    color: ALERT_VARIANT.info
  },
};

export const Success: Story = {
  args: {
    color: ALERT_VARIANT.success
  },
};

export const Warning: Story = {
  args: {
    color: ALERT_VARIANT.warning
  },
};

export const Critical: Story = {
  args: {
    color: ALERT_VARIANT.critical
  },
};

export const Danger: Story = {
  args: {
    color: ALERT_VARIANT.danger
  },
};

export const Neutral: Story = {
  args: {
    color: ALERT_VARIANT.neutral
  },
};

function render(args: Args) {
  return {
    components: { AcvAlert, AcvButton },
    setup: () => ({ args }),
    template: `
      <AcvAlert v-bind="args">
        <template v-if="args.title" #title>${args.title}</template>
        <template v-if="args.description" #description>${args.description}</template>
        <template v-if="args.icon" #icon>${args.icon}</template>
        <template v-if="args.actions" #actions>${args.actions}</template>
        <template v-if="args.right" #right>${args.right}</template>
        ${args.default ? args.default : ''}
      </AcvAlert>`,
  };
}
