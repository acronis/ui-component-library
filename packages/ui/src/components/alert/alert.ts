import type { ExtractPropTypes, PropType } from 'vue';
import type { ConfigurableProps, UikitProps } from '../../utils/props.ts';
import { buildProps } from '../../utils/props.ts';
import type { Expand } from '../../types/generic.ts';

export const ALERT_VARIANT = {
  info: 'info',
  success: 'success',
  warning: 'warning',
  critical: 'critical',
  error: 'error',
  unknown: 'unknown',
} as const;

export type AcvAlertVariant = typeof ALERT_VARIANT[keyof typeof ALERT_VARIANT];

export const AcvAlertIconTypes = {
  [ALERT_VARIANT.info]: 'clr-info--16',
  [ALERT_VARIANT.success]: 'clr-success--16',
  [ALERT_VARIANT.warning]: 'clr-warning--16',
  [ALERT_VARIANT.critical]: 'clr-critical--16',
  [ALERT_VARIANT.error]: 'clr-danger--16',
  [ALERT_VARIANT.unknown]: 'clr-unknown--16'
};

export interface AcvAlertProps {
  /** Descriptive text. Can also be passed with the default slot */
  description?: string

  /** Displays border */
  showBorder?: boolean

  /**
   * Is close icon visible
   */
  showClose?: boolean

  /** If a variant icon is displayed */
  showIcon?: boolean

  /** Title, can also be passed with the slot */
  title?: string

  /** Subtitle, can also be passed with the slot    */
  subtitle?: string

  /** Alert variant */
  color?: AcvAlertVariant
}

export const alertProps: Expand<UikitProps<{ inherit: { default: null, type: BooleanConstructor } } & {
  showClose: BooleanConstructor
  subtitle: StringConstructor
  variant: PropType<AcvAlertVariant>
  description: { type: StringConstructor, required: boolean }
  title: StringConstructor
  showBorder: BooleanConstructor
  showIcon: BooleanConstructor
}>> = buildProps(
  {
    /** Descriptive text. Can also be passed with the default slot */
    description: String,

    /** Displays border */
    showBorder: Boolean,

    /**
     * Is close icon visible
     */
    showClose: Boolean,

    /** If a variant icon is displayed */
    showIcon: Boolean,

    /** Title, can also be passed with the slot */
    title: String,

    /** Subtitle, can also be passed with the slot    */
    subtitle: String,

    /** Alert variant */
    variant: String as PropType<AcvAlertVariant>
  }
);

export interface AcvAlertEvents {
  /**
   * Emitted when the close button is clicked.
   */
  (e: 'close'): void
}

// export type AcvAlertProps = ExtractPropTypes<typeof alertProps>;

export type AlertConfigurableProps = ConfigurableProps<ExtractPropTypes<typeof alertProps>>;

export interface AcvAlertSlots {
  /**
   * Custom content
   */
  default?: void

  /**
   * Title
   * @binding {string} title - content of the title
   */
  title?: string | (() => NonNullable<unknown>)

  /**
   * Subtitle
   * @binding {string} subtitle - content of the subtitle
   */
  description?: string | (() => NonNullable<unknown>)

  /**
   * Content for the AcvAlert icon
   */
  icon?: (() => NonNullable<unknown>)

  /**
   * Content for the actions
   */
  actions?: void

  /**
   * Content for the right block
   */
  right?: () => NonNullable<unknown>
}
