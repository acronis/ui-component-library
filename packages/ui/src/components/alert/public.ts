export { default as AcvAlert } from './alert.vue';
export * from './alert.ts';
