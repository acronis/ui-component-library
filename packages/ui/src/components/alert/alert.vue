<script setup lang="ts">
  import type { AcvAlertSlots, AlertEvents, AlertProps } from './alert.ts';
  import './alert.css';

  defineOptions({
    name: 'Alert',
  });

  const {
    variant,
    title,
    description,
    showClose,
    showIcon
  } = withDefaults(defineProps<AlertProps>(), {
    showClose: false,
    variant: 'info',
  });

  defineEmits<AlertEvents>();
  const slots = defineSlots<AcvAlertSlots>();
</script>

<template>
  <div
    class="acv-alert"
    :class="variant"
    role="alert"
  >
    <div class="content">
      <div
        v-if="showIcon || $slots.icon"
        class="icon"
      >
        <slot name="icon">
          <span v-if="variant === 'info'">I</span>
          <span v-else-if="variant === 'warning'">I</span>
          <span v-else-if="variant === 'error'">I</span>
          <span v-else-if="variant === 'critical'">I</span>
          <span v-else-if="variant === 'unknown'">I</span>
          <span v-else-if="variant === 'success'">I</span>
        </slot>
      </div>
      <div
        v-if="title || $slots.title"
        class="title"
      >
        <slot name="title">
          {{ title }}
        </slot>
      </div>
      <div
        v-if="description || $slots.description"
        class="description"
      >
        <slot name="description">
          {{ description }}
        </slot>
      </div>
      <div
        v-else-if="slots.default"
        class="description"
      >
        <slot></slot>
      </div>
      <div
        v-if="slots.actions"
        class="actions"
      >
        <slot name="actions" />
      </div>
      <div
        v-if="slots.right"
        class="right"
      >
        <slot name="right" />
      </div>
    </div>
    <div
      v-if="showClose"
      class="close"
      @click="$emit('close')"
    >
      X
    </div>
  </div>
</template>

<style scoped>
  .acv-alert {
    display: grid;
    grid-auto-columns: auto 64px;
    border-radius: var(--acv-alert-border-radius);
    column-gap: 24px;
    color: var(--acv-color-text-primary);
    font-family: var(--acv-font-family-default);
    font-size: var(--acv-font-size-body);
    line-height: var(--acv-font-line-height-medium);

    .content {
      display: grid;
      grid-template-columns: min-content 1fr;
      grid-auto-columns: auto;
      grid-template-rows: auto auto auto;
      padding: 8px 0 8px 24px;
    }

    &.info {
      background-color: var(--acv-color-status-info-primary);
    }

    &.success {
      background-color: var(--acv-color-status-success-primary);
    }

    &.warning {
      background-color: var(--acv-color-status-warning-primary);
    }

    &.critical {
      background-color: var(--acv-color-status-critical-primary);
    }

    &.error {
      background-color: var(--acv-color-status-danger-primary);
    }

    &.unknown {
      background-color: var(--acv-color-status-neutral-primary);
    }
  }

  .icon {
    display: flex;
  }

  .title {
    color: var(--acv-alert-title-color);
    grid-column: 2;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;

    /* TODO use mixin */
    font-weight: var(--acv-font-weight-strong);
  }

  .description {
    grid-column: 2;
    word-break: break-word;
  }

  .actions {
    display: flex;
    align-items: center;
    gap: 16px;
    grid-column: 2;
    margin-top: 8px;
  }

  .right {
    display: flex;
    align-items: center;
    grid-row: 1/-1;
    grid-column: 3;
  }

  .close {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 64px;
    height: 100%;
    cursor: pointer;
    margin-left: auto;
    grid-column: 2;
    border-left: var(--acv-border-regular) var(--acv-color-status-info-primary);
  }

  .right ~ .close {
    grid-column: 4;
  }
</style>
