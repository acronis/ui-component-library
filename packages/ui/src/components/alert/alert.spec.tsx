import { describe, expect, it } from 'vitest'
import { mount } from '@vue/test-utils'
import Alert from './alert.vue'
import type { AlertProps } from './alert'
import { AlertIconTypes } from './alert'

describe('alert', () => {
  it('default props', () => {
    const wrapper = mount(Alert)
    expect(wrapper.props()).toMatchInlineSnapshot(`
      {
        "description": undefined,
        "showBorder": false,
        "showClose": false,
        "showIcon": false,
        "subtitle": undefined,
        "title": undefined,
        "variant": "info",
      }
    `)
  })

  it('types', () => {
    ;(['info', 'success', 'warning', 'critical', 'error', 'unknown'] as const).forEach((type) => {
      const wrapper = mount(() => <Alert icon type={type}></Alert>)

      expect(wrapper.find('.acv-alert').classes()).toContain(`${type}`)
      expect(wrapper.findComponent(AlertIconTypes[type]).exists()).toBe(true)
    })
  })

  it('pass props', () => {
    const wrapper = mount(Alert, {
      props: {
        showClose: true,
        variant: 'success',
      } as AlertProps,
    })
    expect(wrapper.props()).toMatchInlineSnapshot(`
      {
        "description": undefined,
        "showBorder": false,
        "showClose": true,
        "showIcon": false,
        "subtitle": undefined,
        "title": undefined,
        "variant": "success",
      }
    `)
  })

  it('renders by default', () => {
    const wrapper = mount(Alert)

    expect(wrapper.element).toMatchInlineSnapshot(`
      <div
        class="acv-alert info"
        data-v-437210c2=""
        role="alert"
      >
        <div
          class="content"
          data-v-437210c2=""
        >
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
          <!--v-if-->
        </div>
        <!--v-if-->
      </div>
    `)
  })

  it('renders title slot', () => {
    const wrapper = mount(Alert, {
      slots: {
        title: 'Title',
      },
    })

    expect(wrapper.get('.title').text()).toBe('Title')
  })

  it('renders description slot', () => {
    const wrapper = mount(Alert, {
      slots: {
        description: 'Description',
      },
    })

    expect(wrapper.text()).toBe('Description')
  })

  it('renders close button', () => {
    const wrapper = mount(Alert, {
      props: {
        showClose: true,
      } as AlertProps,
    })

    expect(wrapper.text()).toBe('X')
  })
})
