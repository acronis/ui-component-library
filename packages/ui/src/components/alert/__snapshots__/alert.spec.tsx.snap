// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`alert > renders close button 1`] = `
<div
  class="close"
  data-v-437210c2=""
>
  <button
    class="acv-button ghost medium primary"
    data-v-437210c2=""
    data-v-7a9642c5=""
    type="button"
  >
    <!--v-if-->
    <!--v-if-->
    <span
      class="content"
      data-v-7a9642c5=""
    >
      
      <svg
        class="acv-icon"
        data-v-437210c2=""
        fill="none"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.295 12.278a1.008 1.008 0 1 0 1.427 1.427L8 9.426l4.278 4.279a1.008 1.008 0 1 0 1.427-1.427L9.426 8l4.279-4.278a1.008 1.008 0 1 0-1.427-1.427L8 6.574 3.722 2.295a1.008 1.008 0 1 0-1.427 1.427L6.574 8l-4.279 4.278Z"
          fill="currentColor"
        />
      </svg>
      
    </span>
    <!--v-if-->
  </button>
</div>
`;
